2 단계 - 문자열 계산기

Jinni01 opened the pull request.

Commit:
feat: Step1 feedback 반영

-  data class 선언시 제공되는 메서드 활용 시도

Commit:
feat: 계산기 클래스 추가 및 입력값 분해 테스트 케이스 작성

Commit:
feat: 공백 문자열 입력 테스트 케이스 작성

Commit:
feat: null 입력 테스트 케이스 작성

Commit:
feat: 더하기 연산 테스트 케이스 작성

Commit:
feat: 빼기 연산 테스트 케이스 작성

Commit:
feat: 곱하기 연산 테스트 케이스 작성

Commit:
feat: 나누기 연산 테스트 케이스 작성

Commit:
feat: 목록에 없는 연산 기호 사용 테스트 케이스 작성

Commit:
feat: 사칙 연산 테스트 케이스 작성

lhs0610 commented on src/main/kotlin/utilities/StringBasedCalculator.kt:
```diff
@@ -0,0 +1,57 @@
+            MINUS.operator -> prefix - postfix
+            MULTIPLY.operator -> prefix * postfix
+            DIVIDE.operator -> prefix / postfix
+            else -> throw IllegalArgumentException("허용되지 않은 사칙연산 기호 \'${operator}\'가 포함되어 있습니다.")
```
친절한 예외 메세지 👍

lhs0610 commented on src/main/kotlin/utilities/StringBasedCalculator.kt:
```diff
@@ -0,0 +1,57 @@
+
+import constant.StringConstant
+import enums.Operators
+import enums.Operators.*
```
import wildcard(`*`)를 지양해주세요

Jinni01 replied:
프로젝트 설정에서 자동변환하지 않도록 변경하겠습니다

lhs0610 commented on src/main/kotlin/utilities/StringBasedCalculator.kt:
```diff
@@ -0,0 +1,57 @@
+import enums.Operators
+import enums.Operators.*
+
+class StringBasedCalculator(
+    expression: String?,
+    var numbers: List<String> = mutableListOf(),
+    var operators: List<String> = mutableListOf()
+) {
```
문자열 계산기는 별도의 상태를 가질 필요가 없는 싱글턴 객체로 설계가 가능할 것 같은데 어떻게 생각하시나요?
수식을 연산하려 할 때마다 인스턴스를 만들 필요는 없을 것 같아요.

Jinni01 replied:
api client같은 느낌과 비슷해서 단일객체 하나만 싱글턴으로 존재해도 될 것 같습니다. 반영해 보겠습니다!

lhs0610 commented on src/main/kotlin/utilities/StringBasedCalculator.kt:
```diff
@@ -0,0 +1,57 @@
+) {
+
+    init {
+        if (expression.isNullOrBlank()) {
+            throw IllegalArgumentException("입력값은 null 이거나 공백 문자일 수 없습니다.")
+        }
```
코틀린에서는 예외를 활용한 코드의 제한이 가능합니다. 
```suggestion
       require (!expression.isNullOrBlank()) {
            throw IllegalArgumentException("입력값은 null 이거나 공백 문자일 수 없습니다.")
        }
```

Jinni01 replied:
require구문이 있군요! 활용해보겠습니다

lhs0610 commented on src/main/kotlin/utilities/StringBasedCalculator.kt:
```diff
@@ -0,0 +1,57 @@
+import enums.Operators.*
+
+class StringBasedCalculator(
+    expression: String?,
+    var numbers: List<String> = mutableListOf(),
+    var operators: List<String> = mutableListOf()
```
추가적으로 가변상태의 프로퍼티를 public하게 제공하면 외부에 의한 위변조가 가능하기에 너무 위험한 코드입니다.

Jinni01 replied:
변수 공개 레벨 관련해서 엄격한 규칙을 적용하도록 하겠습니다

lhs0610 commented on src/main/kotlin/constant/StringConstant.kt:
```diff
@@ -0,0 +1,9 @@
+        const val BLANK = " "
+        const val EMPTY = ""
+    }
+}
```
도메인의 의도를 외부로 노출하는 상수 클래스는 굳이 작성할 필요가 없는 것 같네요

lhs0610 commented on src/main/kotlin/enums/Operators.kt:
```diff
@@ -0,0 +1,16 @@
+                it.operator
+            }.toList()
+        }
+    }
```
내부 연산자를 모두 꺼내 외부에서 연산을 할 필요가 있을까요? 연산자를 차라리 캐싱해서 시간복잡도 O(1)로 찾도록해서 연산하는건 어떨까싶네요

lhs0610 commented on src/main/kotlin/utilities/StringBasedCalculator.kt:
```diff
@@ -0,0 +1,57 @@
+
+    private fun solveExpression(prefix: Int, postfix: Int, operator: String): Int {
+        return when(operator) {
+            PLUS.operator -> prefix + postfix
```
연산의 책임도 직접 가져가기보단 열거타입의 각 인스턴스(PLUS, MINUS, MULTIPLY, )가 가져가도록 해도 좋을 것 같네요.

lhs0610 commented on src/main/kotlin/utilities/StringBasedCalculator.kt:
```diff
@@ -0,0 +1,57 @@
+        val convertedNumbers = numbers.map {
+            val origin = it
+            it.toIntOrNull() ?: throw java.lang.IllegalArgumentException("잘못된 수식 \'${origin}\'이 포함되어 있습니다.")
+        }
```
로직에서 직접 숫자 유효성 검증을하는건 웹 MVC로 치면 트랜잭션 스크립트 패턴으로 보여요. 값 객체를 만들어서 책임을 전담시키는건 어떨까요?

lhs0610 commented on src/main/kotlin/utilities/StringBasedCalculator.kt:
```diff
@@ -0,0 +1,57 @@
+        }
+
+        val numbersIterator = convertedNumbers.iterator()
+        val operatorsIterator = operators.iterator()
```
iterator를 지원하는 데이터 파이프라인 연산자가 많이 제공되고 있습니다. 이를 좀 더 활용해보면 어떨까요?

lhs0610 commented on src/test/kotlin/ComponentTest.kt:
```diff
@@ -0,0 +1,86 @@
+import utilities.StringBasedCalculator
+import kotlin.math.exp
+
+class ComponentTest : ExpectSpec({
```
보통 테스트 클래스는 ${테스트 대상 클래스명}Test 로 짓는게 관례에요. 그게 찾기도 편하니까요.ComponentTest라고 하면 Component라는 클래스에 대한 테스트 겠구나 하는데,  해당 프로덕션 클래스가 없으니 찾기가 힘들죠

Jinni01 replied:
넵 네이밍 컨벤션 맞출 수 있도록 변경하겠습니다

lhs0610 commented on src/test/kotlin/ComponentTest.kt:
```diff
@@ -0,0 +1,86 @@
+
+            val calculator = StringBasedCalculator("1 + 3 - 2 * 4 / 8")
+
+            println(calculator.numbers)
+            println(calculator.operators)
```
검증문에서 출력을 할 필요가 있을까요?

Jinni01 replied:
개발과정중 작성된 출력문들은 제거하도록 하겠습니다

lhs0610 commented on src/test/kotlin/ComponentTest.kt:
```diff
@@ -0,0 +1,86 @@
+            println(calculator.numbers)
+            println(calculator.operators)
+
+            (calculator.numbers.size ?: 0) shouldBe 5
+            (calculator.operators.size ?: 0) shouldBe 4
```
단언문에서 Elvis연산자를 쓰는 것 자체가 로직의 결과를 유추하지 못한다고 보여질 수 있어요. 
```suggestion
            calculator.numbers shouldNotBe null
            calculator.numbers shouldHaveSize 5
```

Jinni01 replied:
제거하도록 하겠습니다

lhs0610 commented on src/test/kotlin/ComponentTest.kt:
```diff
@@ -0,0 +1,86 @@
+
+        expect("잘못된 입력값 테스트: 공백 문자열") {
+
+            val calculator = StringBasedCalculator(StringConstant.EMPTY)
```
이게 무엇을 테스트하려는지 단언문도 없고 의도가 없네요!

Jinni01 replied:
표현식 입력값에 null이나 공백을 삽입하면 에러 메시지가 출력되는데 이 동작에 대한 코드였습니다.
테스트로 더 잘 표현하기 위해서는 어떤 식으로 변경해볼 수 있을까요?

lhs0610 commented on src/test/kotlin/ComponentTest.kt:
```diff
@@ -0,0 +1,86 @@
+            result shouldBe -6
+        }
+
+        expect("나누기 연산 테스트") {
```
0으로 나눌때에 대한 테스트나 0을 숫자로 나눌때에 대한 검증도 있어야 하지 않을까요?

Jinni01 replied:
연산 과정에서 ArithmeticExcetpion이 발생할 수 있겠네요. 방어식을 추가하도록 하겠습니다

lhs0610 reviewed (changes requested):
의진님 2단계 진행 바로 해주셨네요 👍 
태그걸고 요청해달라니까 왜 안해주셨어요 ㅋㅋ 

몇 가지 피드백남겨드렸어요. 추가적으로 코드레벨보단 구조레벨에서 너무 하나의 클래스에 기능들이 몰려있는 느낌을 받았어요. 좀 더 책임에 따라 객체를 나눠봐도 좋을 것 같아요. 

피드백 확인 후 다시 리뷰요청 부탁드려요

Commit:
refactor: Step2 피드백 반영 및 개선

- StringBasedCalculatorV2 추가(개선판, 머지시 기존 클래스 교체)
- 계산기 클래스 인스턴스에서 유틸클래스로 변경(정적메소드 사용)
- 나누기 연산 0 테스트 케이스 추가
- 디버깅용 프린트구문 제거
- if문 예외 체크 require로 변경

catsbi commented on src/main/kotlin/domain/CalculationDto.kt:
```diff
@@ -0,0 +1,38 @@
+import enums.ErrorCode
+import enums.Operator
+
+class CalculationDto(
```
DTO의 의미는 Data Transfer Object에요. 즉 데이터 전달 의외의 목적을 가지지 않는다는 건데, 해당 객체는 DTO로 보이지는 않습니다 🤔

Jinni01 replied:
적합한 변수명으로 변경해볼 수 있도록 하겠습니다

catsbi commented on src/main/kotlin/domain/CalculationDto.kt:
```diff
@@ -0,0 +1,38 @@
+import enums.Operator
+
+class CalculationDto(
+    val numbers: List<Int> = mutableListOf(),
+    val operators: List<String> = mutableListOf()
```
해당 프로퍼티 역시 외부에서 위변조가 충분히 가능할 것 같은데, 어떻게 생각하시나요
```kotlin
val mutableList = mutableListOf()
val calculation = CalculationDto(numbers = mutableList, operators = listOf())

mutableList.add(1)
```

이렇게 코드를 작성한다면 해당 DTO의 내부 값은 충분히 위변조가 가능해보입니다.

Jinni01 replied:
방어적 복사를 수행해야 할 필요가 있다 생각합니다. 개선해보도록 하겠습니다

catsbi commented on src/main/kotlin/domain/CalculationDto.kt:
```diff
@@ -0,0 +1,38 @@
+    companion object {
+
+        fun from(expression: String?): CalculationDto {
+            require(!expression.isNullOrBlank()) { ErrorCode.ERROR_EXPRESSION_NOT_NULL_OR_BLANK.code + ": 입력값은 null 이거나 공백 문자일 수 없습니다." }
```
문자열 + 연산을 할 필요는 없어요. 
```kotlin
"${ERROR_EXPRESSION_NOT_NULL_OR_BLANK.code} : 입력값은 null 이거나 공백 문자일 수 없습니다."
```

catsbi commented on src/main/kotlin/domain/CalculationDto.kt:
```diff
@@ -0,0 +1,38 @@
+        fun from(expression: String?): CalculationDto {
+            require(!expression.isNullOrBlank()) { ErrorCode.ERROR_EXPRESSION_NOT_NULL_OR_BLANK.code + ": 입력값은 null 이거나 공백 문자일 수 없습니다." }
+
+            val split = expression.split(StringConstant.BLANK)
```
외부에 구분자를 의존하는데, 해당 구분자가 바뀌게 된다면 어떻게 될까요? 코드를 수정해야하고, 그렇게 되면 OCP 원칙을 깨트리게 되는데 어떻게 해결할 수 있을까요?

Jinni01 replied:
떠오르는 아이디어가 여러 가지 있어 나열해보았습니다.

1. 내부에서 고정값을 사용하도록 변경(const val 필드로 구분자를 갖는다. const val splitter = " ") -> 외부 의존은 제거되지만 새로운 구분자 사용시 새 분할 로직을 작성해주어야 하는 점은 같음
2. 상위 인터페이스를 선언해 getSpliiter()메서드를 지정한다. 구현을 원하는 구분자를 사용하는 각 클래스를 만든다. (경우에 따라 팩토리도)

생각해 본 것은 이 정돈데, 추가적인 조언 주신다면 감사하겠습니다.

lhs0610 replied:
둘 다 좋은 생각같아요. 의진님이 원하시는 방향으로 해보세요 ㅎㅎ

catsbi commented on src/main/kotlin/domain/CalculationDto.kt:
```diff
@@ -0,0 +1,38 @@
+            }
+
+            val operators = separate[true] ?: mutableListOf()
+            val numbers = separate[false] ?: mutableListOf()
```
1. 연산자를 매번 다 꺼내온 뒤 contains로 확인하는 과정을 줄일 수 없을까요? 
2. "5 5 - 1 / 7 3" 이런식으로 입력값이 전달되면 해당 로직은 제대로 분리를 할 수 없을 것 같네요. 짝수 인덱스와 홀수 인덱스가 각각 연산자/피연산자 역할을 맡고 있는데, 이를 활용해보면 어떨까요

catsbi commented on src/main/kotlin/domain/CalculationDto.kt:
```diff
@@ -0,0 +1,38 @@
+        fun of(numbers: List<String>, operators: List<String>): CalculationDto {
+
+            val convertedNumbers = numbers.map {
+                val origin = it
```
지역 변수를 할당한 이유가 있나요?

Jinni01 replied:
에러메시지에 포함하기 위해 할당해 두었었는데, it을 사용해도 무방한 코드였네요!

catsbi commented on src/main/kotlin/enums/ErrorCode.kt:
```diff
@@ -0,0 +1,9 @@
+package enums
+
+enum class ErrorCode(var code: String) {
+    //input errors
```
불필요한 주석은 제거해주세요

catsbi commented on src/main/kotlin/enums/Operator.kt:
```diff
@@ -1,14 +1,14 @@
     DIVIDE("/");
 
     companion object {
         fun getOperators(): List<String> {
```
매번 리플렉션을 통해 연산자를 모두 외부로 반환해서 계산을 외부에서 수행하고 있어요. 🤔

catsbi commented on src/main/kotlin/utilities/StringBasedCalculatorV2.kt:
```diff
@@ -0,0 +1,45 @@
+import enums.ErrorCode
+import enums.Operator
+
+class StringBasedCalculatorV2 {
```
v2로 만들기보단 기존껄 수정해도 좋을 것 같네요🤔

Jinni01 replied:
앗 요건 커밋메시지에 남기긴 했는데 기존파일 유지 후 대치를 생각했었습니다. 나머지 과정 중 같은 경우 기존클래스 수정하도록 하겠습니다.

catsbi commented on src/main/kotlin/utilities/StringBasedCalculatorV2.kt:
```diff
@@ -0,0 +1,45 @@
+import enums.ErrorCode
+import enums.Operator
+
+class StringBasedCalculatorV2 {
```
해당 객체는 StringBasedCalculatorV2()으로 인스턴스화가 가능할 것 같아요. 싱글턴 패턴을 적용하려면 object 키워드를 사용해보면 좋을 것 같아요.

Jinni01 replied:
유틸리티클래스로 만든다고 생각하다가 싱글턴화를 빠트렸네요. 자바와 코틀린 개념 사이에서 혼동중인 것 같습니다. 
잘 이해하고 사용할 수 있도록 해보겠습니다

catsbi commented on src/main/kotlin/utilities/StringBasedCalculatorV2.kt:
```diff
@@ -0,0 +1,45 @@
+        }
+
+        private fun calculate(calculationDto: CalculationDto): Int {
+            val numbersIterator = calculationDto.numbers.iterator()
+            val operatorsIterator = calculationDto.operators.iterator()
+
+            var result = numbersIterator.next()
+
+            while(numbersIterator.hasNext() && operatorsIterator.hasNext()) {
+                val number = numbersIterator.next()
+                val operator = operatorsIterator.next()
+                result = solveExpression(result, number, operator)
+            }
+
+            return result
```
코틀린은 다양한 컬렉션API를 제공하고 있어요. 이렇게 Iterator를 직접 꺼내서 while로 반복하는건 이제 자바에서도 안하는 방식 아닌가요 ㅋㅋㅋㅋ😆

catsbi reviewed (changes requested):
의진님, V2를 만들면서 새로 시도해본건 좋은데, 추가적으로 피드백 남겨드렸어요. 
아직 코틀린에 익숙하지 않아서, 제대로 활용하지 못하는 부분들이 있는것 같아요. 
확인 후 다시 리뷰요청 부탁드려요

Commit:
refactor: Step2 피드백 반영 및 개선

- 리스트 생성자 인자 방어적 복사
- 에러메시지 문자열 타입 더하기 연산 제거
- dto -> 용도에 맞지 않는 클래스명 변경
- 불필요 지역변수 할당 제거
- StringBasedCalculator 싱글턴 객체로 변경
- 테스트 케이스 추가: 잘못된 수식
- 테스트 케이스 추가: 커스텀 구분자 사용
- 테스트 케이스 추가: 피연산자 위치에 다른 값 입력
- 내부 enum 캐싱하도록 변경
- 사용되지 않는 클래스 Deprecated 처리 (다음커밋시 제거)

lhs0610 commented on src/main/kotlin/enums/Operator.kt:
```diff
@@ -1,16 +1,28 @@
-            return enumValues<Operator>().map {
-                it.operator
-            }.toList()
+        private val mapping = enumValues<Operator>().associate { it.operator to it.calculation }
```
```suggestion
        private val mapping =  values().associateBy(Operator::symbol)
```

associateBy로 좀 더 간편하게 매핑이 가능합니다

Jinni01 replied:
associateBy를 사용하면 객체 자체가 매핑이 되어서 associate로 operator -> 람다 함수 매핑되도록 작성했습니다.

틀린 부분 있다면 설명해주시면 감사하겠습니다 ㅎ.ㅎ

Jinni01 replied:
enumValues() 부분은 values()로 변경하도록 하겠습니다.

lhs0610 commented on src/main/kotlin/enums/Operator.kt:
```diff
@@ -1,16 +1,28 @@
-            }.toList()
+        private val mapping = enumValues<Operator>().associate { it.operator to it.calculation }
+
+        fun getMapping(): Map<String, (Int, Int) -> Int> {
+            return mapping
+        }
```
OOP 설계의 핵심은 객체간 메세지 요청/응답을 통해 대화를 하도록 설계하는 것입니다. 
즉, 내부 값을 외부로 노출시키기보단 내부에 캐싱되어있는 값에서 적절한 값을 찾아서 반환만 해줘도 될 것 같네요. 
```suggestion
        fun getMapping(symbol: String): Operator = mapping[symbol] ?: throw IllegalArgumentException(...)
```

혹은 이펙티브 코틀린에서 제시하는 방식 중 예외를 던지기보단 예외를 적절히 다루라는 내용을 토대로
none enum으로 null반환 혹은 throw Exception을 하기보단, None enum타입을 만들어 제공할 수도 있습니다.

lhs0610 commented on src/main/kotlin/enums/Operator.kt:
```diff
@@ -1,16 +1,28 @@
+            val calculation = mapping[operator]
+
+            return calculation!!(left, right)
         }
```
매개변수가 3개가 되네요. 2개이하로 줄여볼 수 있어요. 
호출하는 측에서 다음과 같이 작성하면 어떨까요? 
```kotlin
acc = Operator.getMapping(symbol).calculation(acc, operand)
```

lhs0610 commented on src/main/kotlin/utilities/StringBasedCalculator.kt:
```diff
@@ -1,56 +1,43 @@
+        var result = convertStringToIntCheckNumeric(split[FIRST_OPERAND_INDEX]);
 
-        var result = numbersIterator.next();
+        for (index in OPERATOR_OFFSET until split.size step 2) {
+            val operator = getExistOperator(split[index])
+            val operand = convertStringToIntCheckNumeric(split[index + OPERATOR_OFFSET])
```
코틀린의 컬렉션 API를 더 활용할수도 있습니다 껄껄껄. 하지만 이대로도 진행하는데는 문제 없을 것 같네요!

Jinni01 replied:
코틀린 컬렉션 API를 제대로 활용하지는 못했다는 느낌을 받았는데, 힌트 주신 내용 참고삼아 더 코틀린스럽게 바꿔볼 수 있도록 하겠습니다.

lhs0610 commented on src/main/kotlin/utilities/StringBasedCalculator.kt:
```diff
@@ -1,56 +1,43 @@
-            else -> throw IllegalArgumentException("허용되지 않은 사칙연산 기호 \'${operator}\'가 포함되어 있습니다.")
-        }
+    private fun convertStringToIntCheckNumeric(origin: String): Int {
+        return origin.toIntOrNull() ?: throw java.lang.IllegalArgumentException("${ErrorCode.ERROR_IS_NOT_NUMERIC.code}: \'${origin}\'은 숫자가 아닙니다.")
```
FQCN으로 표현하고 있는데, 간소화 할 수 있을 것 같네요

Jinni01 replied:
패키지 임포트 설정을 뭔가 잘못 만진 것 같네요. 수정할 수 있도록 하겠습니다

lhs0610 reviewed (approved):
의진님, 피드백이 적지 않은데 다 고민하시느라 고생 많으셨어요. 
몇 가지 피드백 남겨드렸는데 확인해보시고 다음 미션에서 같이 반영해보면 좋을 것 같네요. 
머지하도록 하겠습니다.
